puppetlabs/puppetlabs-dsc
# (maint) Pay down tech debt

The PR is purely maintenance work and introduces no new features, but it does remove an unused task and extracts a lot of code to ruby classes that were previously sitting free flowing inside rake tasks.  This will reduce pain in the future when we modify this to pull DSC Resources from sources other than git.

Commit 7f0876e7 is most important, and could be considered a bugfix all on it's own, as it standardizes the variables that control folder placement inside both the rake task and inside the manager.rb class. Previously these were hardcoded, but gave the impression that they could be separately controlled when in fact it was possible to specify a new path in the rake task but have it ignored by the manager.rb class. 

This PR also renames some global variables and some local variables defined in the dsc namespace in the rake tasks. Previous variable names were confusing and often similarly named with contrary purposes. This rename leans towards more clarity in naming at the expense of being a little verbose. Proper use of these variables is important, as misuse results in a failed build or a corrupted build that still 'succeeds'.

In summation, this PR:

- Extract document building, type cleaning, and type building to separate classes
- Removes unused global variables from rake tasks
- Renames global and local variables in rake task for clarity
- Removes dsc:module:skeleton unused rake task

To test:

Criteria 1:

- Clean checkout
- `bundle exec rake dsc:clean && bundle exec rake dsc:build`
- No files should change

Criteria 2:

- Clean checkout
- Delete dsc_resource_release_tags.yml
- `bundle exec rake dsc:clean && bundle exec rake dsc:build`
- New DSC Resources should be vendored.

*Note:* If PSExecutionPolicy still has issue https://github.com/PowerShell/xPowerShellExecutionPolicy/issues/14, then it will fail parsing that DSC Resource. This is not related to this PR

*Note:* PR https://github.com/puppetlabs/puppetlabs-dsc/pull/352 addresses problems in SqlServerDsc module manifest parsing
